packages/contracts/contracts/L1/ValidatorPool.sol

```diff
@@ -105,72 +93,196 @@ contract ValidatorPool is ReentrancyGuardUpgradeable, Semver {
+    /**
+     * @notice Bond asset corresponding to the given output index. This function is called when submitting output.
+     *         Note that the amount to bond should be at least the minimum bonding amount,
+     *         and the output root must also be given for later output root comparison.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     * @param _amount      Amount of bonding.
+     * @param _expiresAt   Expiration time (in seconds) for bonding.
+     */
+    function bond(
+        uint256 _outputIndex,
+        uint128 _amount,
+        uint128 _expiresAt
+    ) external payable {
+        require(msg.sender == address(L2_ORACLE), "ValidatorPool: sender is not L2OutputOracle");
+        require(_amount > 0, "ValidatorPool: amount cannot be 0");
+        require(_expiresAt > block.timestamp, "ValidatorPool: expiration time should be future");
+
+        Types.Bonding storage bonding = bonds[_outputIndex];
+        require(bonding.expiresAt == 0, "ValidatorPool: exists bonding of given output index");
+
+        // Unbond someone's bonding if available.
+        tryUnbond();
+
+        address submitter = getOutputSubmitter(_outputIndex);
+        _decreaseBalance(submitter, _amount);
+
+        bonding.amount = _amount;
+        bonding.expiresAt = _expiresAt;
+
+        emit Bonded(submitter, _outputIndex, _amount, _expiresAt);
+    }
+
+    /**
+     * @notice Increases the amount in a bonding corresponding to the given output index.
+     *         This function is called when creating challenge.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     */
+    function increaseBond(address _challenger, uint256 _outputIndex) external payable {
+        Types.Bonding storage bonding = bonds[_outputIndex];
+        require(bonding.expiresAt > 0, "ValidatorPool: bonding is not exists");
+
+        _decreaseBalance(_challenger, bonding.amount);
+        bonding.amount = bonding.amount * 2;
+
+        emit Increased(_challenger, _outputIndex, bonding.amount);
+    }
 
-        if (!validators[msg.sender]) {
-            validators[msg.sender] = true;
-            emit NewValidator(msg.sender);
+    /**
+     * @notice Attempt to unbond. Reverts if unbonding is not possible.
+     */
+    function unbond() external nonReentrant {
+        bool released = tryUnbond();
+        require(released, "ValidatorPool: no bonding that can be unbond");
+    }
+
+    /**
+     * @notice Attempts to unbond a bonding corresponding to nextUnbondOutputIndex
+               and returns whether the unbonding was successful.
+     *
+     * @return Whether a bonding has been successfully unbonded.
+     */
+    function tryUnbond() internal returns (bool) {
+        uint256 outputIndex = nextUnbondOutputIndex;
+        address recipient = getOutputSubmitter(outputIndex);
+
+        Types.Bonding storage bonding = bonds[outputIndex];
+        if (block.timestamp >= bonding.expiresAt && bonding.amount > 0) {
```

Comment:
right.

Reply:
When the bond does not exist, it must be checked for existence because `block.timestamp` is always greater than `bonding.expiresAt`. I intended for it to not unbond when there is no bond by checking that `bonding.amount > 0`.

Reply:
Oh I see. Then it makes sense. Good!